src/dbnode/storage/bootstrap/bootstrapper/base.go

```diff
@@ -76,133 +74,261 @@ func (b baseBootstrapper) String() string {
-		b.Can(bootstrap.BootstrapParallel) &&
-		b.next.Can(bootstrap.BootstrapParallel) {
-		// If ranges can be bootstrapped now from the next source then begin attempt now
-		nextAttempted = true
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			nextStatus, nextErr = step.runNextStep(nextRanges)
-		}()
-	}
+		dataRange := currNamespace.DataRunOptions.ShardTimeRanges.MinMaxRange()
+		requestedLogFields := append(logFieldsCopy(logFields), []zapcore.Field{
+			zap.String("namespace", id.String()),
+			zap.Int("numShards", len(currNamespace.Shards)),
+			zap.Duration("dataRange", dataRange),
+		}...)
+		if dataRange > 0 {
+			dataMin, dataMax := currNamespace.DataRunOptions.ShardTimeRanges.MinMax()
+			requestedLogFields = append(requestedLogFields, []zapcore.Field{
+				zap.Time("dataFrom", dataMin),
+				zap.Time("dataTo", dataMax),
+			}...)
+		}
+		if namespace.Metadata.Options().IndexOptions().Enabled() {
+			indexRange := currNamespace.IndexRunOptions.ShardTimeRanges.MinMaxRange()
+			requestedLogFields = append(requestedLogFields, []zapcore.Field{
+				zap.Duration("indexRange", indexRange),
+			}...)
+			if indexRange > 0 {
+				indexMin, indexMax := currNamespace.IndexRunOptions.ShardTimeRanges.MinMax()
+				requestedLogFields = append(requestedLogFields, []zapcore.Field{
+					zap.Time("indexFrom", indexMin),
+					zap.Time("indexTo", indexMax),
+				}...)
+			}
+		}
 
-	min, max := currRanges.MinMax()
-	logFields := []zapcore.Field{
-		zap.String("source", b.name),
-		zap.String("namespace", namespace.ID().String()),
-		zap.Time("from", min),
-		zap.Time("to", max),
-		zap.Duration("range", max.Sub(min)),
-		zap.Int("shards", len(currRanges)),
+		b.log.Info("bootstrap from source for namespace requested", requestedLogFields...)
+
+		dataAvailable, err := b.src.AvailableData(namespace.Metadata,
+			namespace.DataRunOptions.ShardTimeRanges.Copy(),
+			namespace.DataRunOptions.RunOptions)
+		if err != nil {
+			return bootstrap.NamespaceResults{}, err
+		}
+
+		currNamespace.DataRunOptions.ShardTimeRanges = dataAvailable
+
+		// Prepare index if required.
+		if namespace.Metadata.Options().IndexOptions().Enabled() {
+			indexAvailable, err := b.src.AvailableIndex(namespace.Metadata,
+				namespace.DataRunOptions.ShardTimeRanges.Copy(),
+				namespace.DataRunOptions.RunOptions)
+			if err != nil {
+				return bootstrap.NamespaceResults{}, err
+			}
+
+			currNamespace.IndexRunOptions.ShardTimeRanges = indexAvailable
+		}
+
+		// Set the namespace options for the current bootstrapper source.
+		curr.Namespaces.Set(id, currNamespace)
+
+		// Log the metadata about bootstrapping this namespace.
+		dataRange = currNamespace.DataRunOptions.ShardTimeRanges.MinMaxRange()
+		prepareLogFields := append(logFieldsCopy(logFields), []zapcore.Field{
+			zap.String("namespace", id.String()),
+			zap.Int("numShards", len(currNamespace.DataRunOptions.ShardTimeRanges)),
+			zap.Duration("dataRange", dataRange),
+		}...)
+		if dataRange > 0 {
```

Comment:
Is it worth explicitly logging the case where the range is 0 here after checking available shards? Although I guess that's inferrable from the `dataRange` field

Reply:
It was printing weird negative numbers (since `time.Time` when zero is really old).